### PR TITLE
Allow WS baseUrl in EthereumConnection

### PR DESCRIPTION
**Pull request by willclarktech**
Opening now to get feedback on the direction before I tidy up/complete tests etc.

Step 1 of #853

### CODE REVIEW

File: packages/iov-ethereum/src/wsjsonrpcclient.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  public constructor(socket: StreamingSocket) {
+    this.socket = socket;
+  }
+
+  public async run(request: JsonRpcRequest): Promise<JsonRpcResponse> {
+    await this.socket.connected;
+
+    const promise = new Promise<JsonRpcResponse>((resolve, reject) => {
+      this.socket.events
+        .map(event => parseJsonRpcResponse2(JSON.parse(event.data)))
```
Comment (webmaster128) [inline packages/iov-ethereum/src/wsjsonrpcclient.ts]: Parsing exceptions in here should be ignored. I suggest converting exception to `null` in the map body and then use a null filter like this one: https://github.com/iov-one/ponferrada/blob/95ec319/packages/sanes-chrome-extension/src/extension/background/model/persona/index.ts#L12-L14

File: packages/iov-ethereum/src/wsjsonrpcclient.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        .filter(response => response.id === request.id)
+        .take(1)
+        .addListener({
+          next: resolve,
+          error: reject,
+        });
```
Comment (webmaster128) [inline packages/iov-ethereum/src/wsjsonrpcclient.ts]: Nice streaming flow

File: packages/iov-ethereum/src/wsjsonrpcclient.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  private readonly socket: StreamingSocket;
+
+  public constructor(socket: StreamingSocket) {
+    this.socket = socket;
+  }
+
+  public async run(request: JsonRpcRequest): Promise<JsonRpcResponse> {
+    await this.socket.connected;
+
+    const promise = new Promise<JsonRpcResponse>((resolve, reject) => {
```
Comment (webmaster128) [inline packages/iov-ethereum/src/wsjsonrpcclient.ts]: `const pendingResponse =`?

File: packages/iov-ethereum/src/ethereumconnection.ts [hunk 4]
```diff
@@ -119,7 +120,7 @@ export class EthereumConnection implements AtomicSwapConnection {
   }
 
   private readonly pollIntervalMs: number;
-  private readonly rpcClient: HttpJsonRpcClient;
+  private readonly rpcClient: HttpJsonRpcClient | WsJsonRpcClient;
```
Comment (webmaster128) [inline packages/iov-ethereum/src/ethereumconnection.ts]: Can we have a shared interface for those two?

Comment (willclarktech) [reply packages/iov-ethereum/src/ethereumconnection.ts]: Definitely, this is the main thing I want to refactor in the current approach.

**Comment by willclarktech** [comment]
I've added an extra test for the ws client in many tests for `EthereumConnection`, but I left them out in some of the more involved cases. I'm not entirely happy about this pattern; a better approach is probably to add unit tests for the various JSON RCP clients and then it shouldn't matter which one we use in the `EthereumConnection` tests. That might be beyond the scope of this ticket however.

File: packages/iov-ethereum/src/wsjsonrpcclient.ts (hunk 2 above)
Comment (willclarktech) [reply packages/iov-ethereum/src/wsjsonrpcclient.ts]: ❤️ xstream

File: packages/iov-ethereum/src/ethereumconnection.ts [hunk 5]
```diff
@@ -119,18 +120,24 @@ export class EthereumConnection implements AtomicSwapConnection {
...
   private readonly erc20Tokens: ReadonlyMap<TokenTicker, Erc20Options>;
   private readonly erc20ContractReaders: ReadonlyMap<TokenTicker, Erc20Reader>;
   private readonly codec: EthereumCodec;
 
   public constructor(baseUrl: string, chainId: ChainId, options: EthereumConnectionOptions) {
+    const baseUrlIsHttp = ["http://", "https://"].some(prefix => baseUrl.startsWith(prefix));
+    const baseUrlIsWs = ["ws://", "wss://"].some(prefix => baseUrl.startsWith(prefix));
+    const wsUrl = baseUrlIsWs ? baseUrl : options.wsUrl;
+    if (!wsUrl && !baseUrlIsHttp) {
+      throw new Error("Unsupported protocol for baseUrl: must be one of http, https, ws or wss");
```
Comment (webmaster128) [inline packages/iov-ethereum/src/ethereumconnection.ts]: I think this error messge should be based on `baseUrlIsWs`/`baseUrlIsHttp` only, so `wsUrl` can be set below this block.

File: packages/iov-ethereum/src/jsonrpcclient.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { Stream } from "xstream";
+
+import { JsonRpcRequest, JsonRpcResponse } from "@iov/jsonrpc";
+import { SocketWrapperMessageEvent } from "@iov/socket";
+
+export interface JsonRpcClient {
```
Comment (webmaster128) [inline packages/iov-ethereum/src/jsonrpcclient.ts]: Can we call this `EthereumRpcClient`? We have JsonRpcClient already in `@iov/jsonrpc`

**Comment by webmaster128** [comment]
The CI is most likely because of the `async function loadChainId(baseUrl: string): Promise<ChainId> {` implementation, which uses a HTTP client on a WS URL

**Comment by willclarktech** [comment]
Hm, weird that it passes in Safari though?

**Comment by webmaster128** [comment]
> Hm, weird that it passes in Safari though?

axios, which is used in the HTTP client, is only a wrapper around native APIs. Those APIs can behave differently. However, I did not look into the details.

**Pull request merged**